sklearn/cluster/tests/test_optics.py

```diff
@@ -160,15 +160,16 @@ def test_cluster_hierarchy_(global_dtype):
     "metric, is_sparse",
     [["minkowski", False], ["euclidean", True]],
 )
-def test_correct_number_of_clusters(metric, is_sparse):
+@pytest.mark.parametrize("csr_container", CSR_CONTAINERS)
+def test_correct_number_of_clusters(metric, is_sparse, csr_container):
```

Comment:
To avoid running the `is_sparse=False` solution as many time we have the `csr_container`, we can use the following parametrization:

```python
@pytest.mark.parametrize(
    "csr_container, metric",
    [(None, "minkowski")]
    + [(container, "euclidean") for container in CSR_CONTAINERS],
)
```

Concretely, it avoids running twice the exact same test (for dense) for the `"minkowski"`.